### ISSUE TITLE
Current broadcasting rules are inconsistent

### ISSUE THREAD

**Issue by francisco-dlp** [hyperspy/hyperspy#866]
For example:

```python

>>> s = hs.signals.Signal(np.arange(10))
>>> im = hs.signals.Image(np.ones((32, 32)))
>>> s * im
<Signal, title: , dimensions: (32, 32|10)>
>>> s2 = hs.signals.Signal(np.arange(32))
>>> s2 * im
<Signal, title: , dimensions: (32|32)>

```

This is a bug.

**Comment by vidartf** [comment]
Just for clarity, what would be the expected behavior?

**Comment by francisco-dlp** [comment]
The problem with the example is that the size of the spectrum changes the dimensions of the outcome. This bug was introduced in 0.6 (!)

**Comment by vidartf** [comment]
But what is the expected behavior of `s * im`? A `ValueError`?

**Comment by francisco-dlp** [comment]
No, the wrong behaviour is the expected behaviour according to the [current broadcasting rule](http://hyperspy.org/hyperspy-doc/current/user_guide/tools.html#signal-operations), the problem is that the current broadcasting rules make no sense whatsoever (I wrote them, so I can be this harsh to myself). Have a look at the new broadcasting rules that this PR proposes in `user_guide/tools.rst`

**Comment by vidartf** [comment]
To reformulate the question: How is `s * im` handled with the new rules? As far as I can tell from the new documentation, it is indeed to throw a `ValueError` (?).

I assumed the current behavior was different from what was expected since this was classified as a bug.

**Comment by francisco-dlp** [comment]
You're right, the broadcasting codes wasn't buggy AFAIK, but the broadcasting rules were ;)

With the new broadcasting rules the outcome of `s * im` depends on their dimensions, just like in numpy. If the dimensions are compatible it returns the result, otherwise it raises a `ValueError`. The only difference with numpy is that HyperSpy broadcasts the navigation and signal axes independently.

**Comment by vidartf** [comment]
I was referring to `s * im` from your example (`<|32, 32> * <|10>`), which I think now will become invalid (`ValueError`). However, if one of those signals were converted to be navigation only, it would be ok? I.e.:
```python
>>> im.axes_manager.set_signal_dimensions(2)
>>> s * im
<Signal, title: , dimensions: (32, 32|10)>
```
Yes?

**Comment by francisco-dlp** [comment]
Yes, that's right.

**Comment by francisco-dlp** [comment]
Fixed in #867

**Comment by vidartf** [comment]
I'm okay with this going into 0.8.3, but I think we should double check next week that this didn't inadvertently break anything that relied on the old behavior (I can't think of a robust regex to quickly check this, so maybe manually check code that is not listed as covered by `coverage`?).
